cellmlmanip/model.py

```diff
@@ -619,6 +626,274 @@ def variables(self):
+        return {'variable': new_deriv_variable, 'expression': eqn.args[0]}
+
+    def _convert_state_variable_deriv(self, original_variable, new_variable, cf):
+        """
+        Create relevant variables/equations when converting a state variable.
+        :param original_variable: the variable to be converted [old units]
+        :param new_variable: the new variable representing the converted symbol [new units]
+        :param cf: conversion factor for unit conversion [new units/old units]
+        :return: a dictionary containing the 'variable' and 'expression' for new derivative
+        """
+        # 1. find the derivative equation for this variable
+        # and get the free variable (wrt_variable)
+        eqn = None
+        for equation in self.equations:
+            if equation.args[0].is_Derivative:
+                if equation.args[0].args[0] == original_variable:
+                    eqn = equation
+                    break
+
+        # get free variable symbol
+        # units [x]
+        wrt_variable = eqn.args[0].args[1]
+
+        # 1. create a new variable/equation for original derivative
+        # will have units [old units/x]
+        new_deriv_variable = self._create_new_deriv_variable_and_equation(eqn, original_variable)
+
+        # 2. add a new derivative equation
+        # dnewvar/dx [new units/x] = new_deriv_var [old units/x] * cf [new units/old units]
+        expression = sympy.Eq(sympy.Derivative(new_variable, wrt_variable), new_deriv_variable * cf)
+        self.add_equation(expression)
+        return {'variable': new_deriv_variable, 'expression': eqn.args[0]}
+
+    def _convert_variable_instance(self, original_variable, cf, units, direction):
+        """
+        Internal function to create new variable and an equation for it.
+        :param original_variable: VariableDummy object to be converted [old units]
+        :param cf: conversion factor [new units/old units]
+        :param units: Unit object for new units
+        :param is_output: boolean are we adding input or output
+        :return: the new variable created [new units]
+        """
+        # 1. get unique name for new variable
+        new_name = self._get_unique_name(original_variable.name + '_converted')
+
+        # 2. if original has initial_value calculate new initial value
+        new_value = None
+        if original_variable.initial_value:
+            new_value = original_variable.initial_value * cf
+
+        # 3. copy cmeta_id from original and remove from original
+        new_variable = self.add_variable(name=new_name,
+                                         units=units,
+                                         initial_value=new_value,
+                                         cmeta_id=original_variable.cmeta_id)
+        original_variable.cmeta_id = ''
+
+        # 4. check whether we had an eqn for original
+        # if so, remove it and replace with new_var [new units] = rhs [old units] * cf [new units/old units]
```

Comment:
RHS if input, original_variable if output?

Reply:
Actually, this seems a slightly backwards way of doing the logic in sections 4 & 5. Would it be clearer to have at the outer level a switch on direction:

* If direction is OUTPUT, always just add an equation `new_variable = original_variable * cf`. No need to loop.
* If direction is INPUT, iterate over equations to find the original definition and replace with `new_variable = original_rhs * cf`, then add `original_variable = new_variable / cf`

Same effect overall, but the control is inverted and seems simpler to me. I had to read this three times to convince myself it worked.